src/main/java/com/redhat/devtools/lsp4ij/features/workspaceSymbol/AbstractLSPWorkspaceSymbolContributor.java

```diff
@@ -1,109 +1,133 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import static com.redhat.devtools.lsp4ij.internal.CompletableFutures.isDoneNormally;
+import static com.redhat.devtools.lsp4ij.internal.CompletableFutures.waitUntilDone;
+
+/**
+ * Abstract base class for LSP goto symbol contributors
+ */
+abstract class AbstractLSPWorkspaceSymbolContributor implements ChooseByNameContributorEx {
+
+    private final static Logger LOGGER = LoggerFactory.getLogger(AbstractLSPWorkspaceSymbolContributor.class);
+
+    private final LSPWorkspaceRequestedSymbolTypes requestedSymbolTypes;
+
+    protected AbstractLSPWorkspaceSymbolContributor(@NotNull LSPWorkspaceRequestedSymbolTypes requestedSymbolTypes) {
+        this.requestedSymbolTypes = requestedSymbolTypes;
+    }
+
+    @Override
+    public void processNames(@NotNull Processor<? super String> processor,
+                             @NotNull GlobalSearchScope scope,
+                             @Nullable IdFilter filter) {
+        Project project = scope.getProject();
+        if (project == null) {
+            return;
+        }
+        String queryString = project.getUserData(ChooseByNamePopup.CURRENT_SEARCH_PATTERN);
+        if (queryString == null) {
+            queryString = "";
+        }
+        List<WorkspaceSymbolData> items = getWorkspaceSymbols(queryString, true, project);
+        if (items != null) {
+            items.stream()
+                    .filter(data -> scope.accept(data.getFile()))
+                    .map(NavigationItem::getName)
+                    .forEach(processor::process);
+        }
+    }
+
+    @Override
+    public void processElementsWithName(@NotNull String name,
+                                        @NotNull Processor<? super NavigationItem> processor,
+                                        @NotNull FindSymbolParameters parameters) {
+        List<WorkspaceSymbolData> items = getWorkspaceSymbols(name, false, parameters.getProject());
+        if (items != null) {
+            items
+                    .stream()
+                    .filter(ni -> parameters.getSearchScope().accept(ni.getFile()))
+                    .forEach(processor::process);
+        }
+    }
+
+    private List<WorkspaceSymbolData> getWorkspaceSymbols(@NotNull String name, boolean cancel, Project project) {
+        // Consume LSP 'workspace/symbol' request
+        LSPWorkspaceSymbolSupport workspaceSymbolSupport = LSPWorkspaceSupport.getSupport(project).getWorkspaceSymbolSupport();
+        if (cancel) {
+            workspaceSymbolSupport.cancel();
+        }
+        WorkspaceSymbolParams params = new LSPWorkspaceSymbolParams(name, requestedSymbolTypes);
```

Comment:
This is how we convey the types of symbols being requested from here to `LSPWorkspaceSymbolSupport`. We construct a specialized subclass of the standard `WorkspaceSymbolParams` that contains both the query and the types of symbols being requested. The latter is used to filter language servers based on whether or not they can efficiently support Go To Class-style invocation.